### ISSUE TITLE
Current status?

### ISSUE THREAD

**Issue by jasnell** [WICG/raw-sockets#35]
What is the current status of this proposal? Is there still an intent on moving it forward?

**Comment by phoglenix** [comment]
The project is partially implemented and still intended to go ahead, but is on hold awaiting a more-secure context than is currently available. Work is happening on that front but I don't have an ETA for you, sorry.

**Comment by jasnell** [comment]
Ok. The reason I ask is that I'm currently looking at designing and implementing an overhauled, modernized socket API in Node.js with an eye towards implementing that also in Workers and potentially Deno also. It would be ideal to collaborate on a combined spec but I do have some fairly significant design differences from what's captured in this spec currently. It would definitely be ideal if we could collaborate, especially if there is intent to ship support in browsers.

**Comment by phoglenix** [comment]
Reopening just so other people can see this instead of starting new issues.

We do still intend to get this shipped eventually, but I don't really expect it to happen in 2021. If you have other feedback on the API, please do file issues about it so we can track it when we come back to this .

**Comment by yume-chan** [comment]
Can I try it now in Chromium? With some command line arguments?

**Comment by phoglenix** [comment]
I believe you can test it using the `--enable-blink-features=DirectSockets  --enable-features=DirectSockets --restricted-api-origins=https://www.foo.com,https://www.bar.com` command-line arguments on non-managed devices.

**Comment by yume-chan** [comment]
I figured it out. In command line arguments I only need `--enable-features=DirectSockets --restricted-api-origins=https://www.example.com`, but the HTTP response for the page must include `Cross-Origin-Opener-Policy: same-origin` and `Cross-Origin-Embedder-Policy: require-corp` headers.

**Comment by blackyale** [comment]
@yume-chan Did it work for you? whenever I try this in the canary channel I get "You are using unsupported command line flag: --restricted-api-origins...".

**Comment by yume-chan** [comment]
> Did it work for you? 

Yes, just tested on version 99.0.4778.0 (Official Build) canary (x86_64).

**Comment by phoglenix** [comment]
restricted-api-origins was added in M99. Before M99 it shouldn't be needed

**Comment by blackyale** [comment]
> > Did it work for you?
> 
> Yes, just tested on version 99.0.4778.0 (Official Build) canary (x86_64).

Tried but I can't see the proper methods in 'navigator', am I missing something?
Version 99.0.4797.0 (Official Build) canary (32-bit)

**Comment by yume-chan** [comment]
> I can't see the proper methods in 'navigator'

Did you add those two headers?

**Comment by blackyale** [comment]
> > I can't see the proper methods in 'navigator'
> 
> Did you add those two headers?

Ok just working after adding the headers.

**Comment by komagen084** [comment]
It seems that "openTCPSocket()" not added to Navigator yet.

I use "google chrome 100.0.4860.0 (Official Build) canary （64 bit） (cohort: Clang-64) ".
And, using the "--enable-blink-features=DirectSockets --enable-features=DirectSockets --restricted-api-origins=https://hoge"
But It says 
>Uncaught TypeError: navigator.openTCPSocket is not a function

How can I use "openTCPSocket()"?

**Comment by phoglenix** [comment]
[See yume-chan's response above](https://github.com/WICG/direct-sockets/issues/35#issuecomment-992120727)

**Comment by komagen084** [comment]
Thank you for your answer.
I understood !!
I didn't understand bellow:
First I have to access to the WEB server that response with 'Cross-Origin-Opener-Policy: same-origin' and 'Cross-Origin-Embedder-Policy: require-corp' response header, before I use navigator.openTCPsocket().

**Comment by jimmywarting** [comment]
think it seems very hard to try to enable and test out this new features in chrome. can't get openTCPSocket to show up...

**Comment by enverbisevac** [comment]
this api is very useful especially working with printer port 9100 if it would be possible to work on the local intranet without any restrictions and permissions it would be really great

**Comment by yume-chan** [comment]
The requirements have changed in mid-March (not sure exact version number)

1. `--restricted-api-origins=https://www.example.com` flag
2. `Cross-Origin-Opener-Policy: same-origin` and `Cross-Origin-Embedder-Policy: require-corp` response headers
3. Create a Web App Manifest, add `"isolated_storage": true` in it (https://github.com/robbiemc/pwa-isolated-storage/blob/main/explainer.md)
4. Let users install your page as an app
5. Let users restart your web app

Then `navigator.openTCPSocket` will come back.

![image](https://user-images.githubusercontent.com/1330321/160979829-32aa271a-274d-47b2-8a56-a5a2a27939ec.png)

**Comment by jimmywarting** [comment]
> 4. Let users install your page as an app
> 5. Let users restart your web app

forcing the user to have to install the app in order to use the service feels a tiny bit reluctant

**Comment by robertsdotpm** [comment]
Just giving feedback -- I'm not an expert on browsers like you all. But I second that having to install an app kind of kills this proposal. If it's not as frictionless as the web then it won't be competitive and I believe this API could really improve the web. There are naysayers that are saying it could be too harmful. But you can say the same thing about XSS already. A basic XSS vulnerability in a banking website can already be used to transfer fund. There are obviously impacts of having socket access but I believe they can be mitigated.

# Some suggestions:

- You could block access to TCP 80 and 443 since the browser already has AJAX.
- A HTTP header could be provided by a server to white list hosts (hopefully with good subnet masks / wildcards / patterns.) This would limit XSS to only DDoSing resources you control. I think this is a better approach than having APIs available to the client by default with no restriction. The server would choose what resources it needs and it wouldn't need a prompt from the user.
- For P2P use-cases wild card subnet should be supported. We could make this less damaging by having a black list of IPs like government and military addresses as well as allowing networks to officially black list themselves from connections (I think this is pretty reasonable.)
- Iframes -- maybe just disable the API here, too much potential for abuse.
- Eval -- maybe just disable this when the HTTP header is sent. Helps prevent advertisers doing XSS
- DNS in white list means controlling the DNS lets you control the white list. Maybe not desirable.
- There should be an option to bind the socket beforehand to an arbitrary port. There are usages for this in P2P networking -- such as TCP hole punching which is an esoteric NAT traversal technique.

# Resource control:

Mozilla wrote here = https://github.com/mozilla/standards-positions/issues/431.

Mozilla raised a good point about local resources. Since local addresses are relative to the network you're on specifying them in a white list could give incorrect access to your own local network resources.

I think this would be a good place to prompt for a permission dialogue. In the past, you would only allow access to such resources if they were running as 'trusted' application installed / downloaded to the host. 

# Peer-to-peer:

I think P2P presents one of the most innovative and exciting uses of this API and should be a supported focus. But doing this would require extending the API to add a listen method.

- Being able to control the listen queue number here would be very useful.
- There could be a traffic priority for servers that let you set how bandwidth is used. E.g. use all available bandwidth, use LEDBAT for idle bandwidth, use N%, use absolute value.
- I think you should be able to accept whoever you like.
- I don't think we should worry about botnets. You can already use existing web technology to make a browser-based botnet. Web pages also don't have persistence (unless you're 1337.)
- We should eventually support multi-cast, too, with event listeners. Multi-cast is very important in P2P networking and service discovery use-cases.

# Other thoughts:

I personally don't think we should worry about throttling the traffic of the sockets. But you could build a traffic policy into the HTTP header if you liked. There are ways to do networking that use 'idle bandwidth' or are more 'respectful' of resources (e.g. LEDBAT.) As for repeat connections -- this does seem sketchy but plenty of legitimate network tools would be making many network connections. I think it should be up to the developer to control this and we shouldn't assume bad faith on the part of the website.

In the end there is no cure for simple web attacks and social engineering. A social engineer can still make a website and do a lot more damage than flooding some traffic over the Internet. Besides -- you can already do these attacks anyway with iframes and ajax, right? So I don't really see what sense it would make limiting connection count or neutering this fledgling API.

**Comment by yume-chan** [comment]
> block access to TCP 80 and 443

443 is already blocked.

> limit XSS to only DDoSing resources you control

DDoS is not possible because opening sockets requires transient activation and connect frequency is already limited.

> P2P use-cases

P2P is not possible because Direct Sockets doesn't provide socket server, only client.

Direct Sockets are designed for connecting to legacy system which can't be converted to Web services. P2P is already possible with WebRTC, for example there is https://github.com/webtorrent/webtorrent. WebRTC also provides hole-punching out-of-box.

> Iframes 

Direct Sockets in IFrames are controlled by `Permissions-Policy` header and are blocked by default.

> DNS 

Domain names resolve to local IP addresses are already blocked to mitigate DNS rebind attack.

> But I second that having to install an app kind of kills this proposal.

I don't know if anyone from Chrome team is looking at this issue. So https://bugs.chromium.org/p/chromium/issues/detail?id=1206150 might be a better place to post your thoughts.

**Comment by olalonde** [comment]
Ping. Any status update? Will this spec go forward?

**Comment by phoglenix** [comment]
This work is still paused. The "more-secure context than is currently available" it was waiting on is under active development (https://chromestatus.com/feature/5146307550248960). We will probably reconsider the priority of this work once that is done.

**Comment by robertsdotpm** [comment]
I'm still watching this with great interest and do think it could radically change the web as a delivery platform.

**Comment by tomayac** [comment]
People interested in playing with the API _now_ can see https://github.com/WICG/direct-sockets/issues/46#issuecomment-1436664185.
